File: app/src/main/java/com/testeandroidv2/loginScreen/LoginInteractor.java
package com.testeandroidv2.loginScreen;

import android.util.Log;

interface LoginInteractorInput {
    public void fetchLoginData(LoginRequest request);
}


public class LoginInteractor implements LoginInteractorInput {

    public static String TAG = LoginInteractor.class.getSimpleName();
    public LoginPresenterInput output;
    public LoginWorkerInput aLoginWorkerInput;

    public LoginWorkerInput getLoginWorkerInput() {
        if (aLoginWorkerInput == null) return new LoginWorker();
        return aLoginWorkerInput;
    }

    public void setLoginWorkerInput(LoginWorkerInput aLoginWorkerInput) {
        this.aLoginWorkerInput = aLoginWorkerInput;
    }

    @Override
    public void fetchLoginData(LoginRequest request) {
        aLoginWorkerInput = getLoginWorkerInput();
        LoginResponse LoginResponse = new LoginResponse();
        // Call the workers

        output.presentLoginData(LoginResponse);
    }
}

File: app/src/main/java/dev/vitorpacheco/solutis/bankapp/loginScreen/LoginInteractor.kt
package dev.vitorpacheco.solutis.bankapp.loginScreen

import com.google.common.base.Strings
import dev.vitorpacheco.solutis.bankapp.BankApp.Companion.LAST_LOGGED_USER_KEY
import dev.vitorpacheco.solutis.bankapp.workers.SharedPreferencesRequest
import dev.vitorpacheco.solutis.bankapp.workers.SharedPreferencesWorker

interface LoginInteractorInput {
    fun doLogin(request: LoginRequest)
    fun getLastLoggedUser(request: SharedPreferencesRequest)
}

class LoginInteractor : LoginInteractorInput {

    var output: LoginPresenterInput? = null
    var loginWorker = LoginWorker()
    var sharedPreferencesWorker = SharedPreferencesWorker()

    override fun doLogin(request: LoginRequest) {
        if (Strings.isNullOrEmpty(request.user)) {
            output?.presentLoginData(
                LoginResponse(
                    error = UserError(
                        "Campo obrigatório",
                        UserFormFields.USER
                    )
                )
            )
            return
        }

        if (Strings.isNullOrEmpty(request.password)) {
            output?.presentLoginData(
                LoginResponse(
                    error = UserError(
                        "Campo obrigatório",
                        UserFormFields.PASSWORD
                    )
                )
            )
            return
        }

        loginWorker.login(request) { response ->
            request.context?.let { context ->
                sharedPreferencesWorker.save(SharedPreferencesRequest(context, LAST_LOGGED_USER_KEY, request.user))
            }

            output?.presentLoginData(response)
        }
    }

    override fun getLastLoggedUser(request: SharedPreferencesRequest) {
        output?.presentLoginData(LoginResponse(user = sharedPreferencesWorker.get(request)))
    }

    companion object {
        var TAG = LoginInteractor::class.java.simpleName
    }

}
